Sources/SwiftGRPCNIO/CallHandlers/BaseCallHandler.swift

```diff
@@ -8,29 +8,79 @@ import NIOHTTP1
 public class BaseCallHandler<RequestMessage: Message, ResponseMessage: Message>: GRPCCallHandler {
   public func makeGRPCServerCodec() -> ChannelHandler { return GRPCServerCodec<RequestMessage, ResponseMessage>() }
-  
+
   /// Called whenever a message has been received.
   ///
   /// Overridden by subclasses.
-  public func processMessage(_ message: RequestMessage) {
+  public func processMessage(_ message: RequestMessage) throws {
     fatalError("needs to be overridden")
   }
-  
+
   /// Called when the client has half-closed the stream, indicating that they won't send any further data.
   ///
   /// Overridden by subclasses if the "end-of-stream" event is relevant.
   public func endOfStreamReceived() { }
+
+  /// Whether this handler can still write messages to the client.
+  private var serverCanWrite = true
+
+  /// Called for each error recieved in `errorCaught(ctx:error:)`.
+  private weak var errorDelegate: ServerErrorDelegate?
+
+  public init(errorDelegate: ServerErrorDelegate? = nil) {
+    self.errorDelegate = errorDelegate
+  }
 }
 
 extension BaseCallHandler: ChannelInboundHandler {
   public typealias InboundIn = GRPCServerRequestPart<RequestMessage>
-  public typealias OutboundOut = GRPCServerResponsePart<ResponseMessage>
+
+  /// Passes errors to the user-provided `errorHandler`. After an error has been received an
+  /// appropriate status is written. Errors which don't conform to `GRPCStatusTransformable`
+  /// return a status with code `.internalError`.
+  public func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+    errorDelegate?.observe(error)
+
+    let transformed = errorDelegate?.transform(error) ?? error
+    let status = (transformed as? GRPCStatusTransformable)?.asGRPCStatus() ?? GRPCStatus.processingError
+    self.write(ctx: ctx, data: NIOAny(GRPCServerResponsePart<ResponseMessage>.status(status)), promise: nil)
+  }
 
   public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
     switch self.unwrapInboundIn(data) {
-    case .head: preconditionFailure("should not have received headers")
-    case .message(let message): processMessage(message)
-    case .end: endOfStreamReceived()
+    case .head:
+      // Head should have been handled by `GRPCChannelHandler`.
+      self.errorCaught(ctx: ctx, error: GRPCStatus(code: .unknown, message: "unexpectedly received head"))
+
+    case .message(let message):
+      do {
+        try processMessage(message)
+      } catch {
+        self.errorCaught(ctx: ctx, error: error)
+      }
+
+    case .end:
+      endOfStreamReceived()
+    }
+  }
+}
+
+extension BaseCallHandler: ChannelOutboundHandler {
+  public typealias OutboundIn = GRPCServerResponsePart<ResponseMessage>
+  public typealias OutboundOut = GRPCServerResponsePart<ResponseMessage>
+
+  public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+    guard serverCanWrite else {
+      promise?.fail(error: GRPCStatus.processingError)
```

Comment:
Should we have a different kind of error for this (maybe not even of type `GRPCStatus`), e.g. "attempted to write more than one status"? (See below for a more detailed comment in a similar vein.)

Reply:
Absolutely, this will never actually reach the caller so doesn't even need to be a `GRPCStatus`.